Make RRef::toHere() return a jit::Future

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28025 Make RRef::LocalValue return Future
* **#27943 Make RRef::toHere() return a jit::Future**

This is step 1 to make PyRRef::toHere() non-blocking on caller.

Differential Revision: [D17936747](https://our.internmc.facebook.com/intern/diff/D17936747)

mrshenli committed:
Make RRef::toHere() return a FutureMessage

This is step 1 to make PyRRef::toHere() non-blocking on caller.

[ghstack-poisoned]

mrshenli commented:
Don't merge before #27789 and #27251

mrshenli committed:
Update on "[WIP] Make RRef::toHere() return a FutureMessage"


This is step 1 to make PyRRef::toHere() non-blocking on caller.

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Make RRef::toHere() return a FutureMessage"


This is step 1 to make PyRRef::toHere() non-blocking on caller.

[ghstack-poisoned]

mrshenli committed:
Update on "Make RRef::toHere() return a jit::Future"


This is step 1 to make PyRRef::toHere() non-blocking on caller.

Differential Revision: [D17936747](https://our.internmc.facebook.com/intern/diff/D17936747)

[ghstack-poisoned]

xush6528 commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -101,8 +101,8 @@ std::unique_ptr<RpcCommandBase> RequestCallbackImpl::processRpc(
           ctx.getOrCreateOwnerRRef<IValue>(srf.rrefId());
-      return c10::guts::make_unique<RRefFetchRet>(
-          RRefFetchRet({rref->getValue()}));
+      std::vector<IValue> ivalues = {rref->getValue()};
```
Are you just making the copy explicitly because the line was too long?

mrshenli replied:
After addressing conflicts with master, this code no long exists. :)

mrshenli committed:
Update on "Make RRef::toHere() return a jit::Future"


This is step 1 to make PyRRef::toHere() non-blocking on caller.

Differential Revision: [D17936747](https://our.internmc.facebook.com/intern/diff/D17936747)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/rref.cpp:
```diff
@@ -131,32 +130,39 @@ const ForkId& UserRRef<T>::forkId() const {
+  auto future = std::make_shared<ivalue::Future>(nullptr);
   auto agent = RpcAgent::getDefaultRpcAgent();
-  std::shared_ptr<FutureMessage> fm = agent->send(
+  auto fm = agent->send(
```
When the `ScriptRRefFetchCall`/`PythonRRefFetchCall` arrives, the owner might not exist yet in the callee. So, when handling the fetch all, it could be the first time the callee see that `RRefId` and will need to create a new `OwnerRRef`, which requires to know the type. 

Another option is that, instead of blocking on `OwnerRRef::getValue()`, request handler it can block (or add callback) on `RRefContext` until the `OwnerRRef` becomes ready. In that way, we don't need two messages any more. We can add that in a follow up PR.

xush6528 replied:
I see. In summary, sometimes users know the types before owners. 
For example, a user asks to run a user function. The return value will be returned as an RRef. The type of the return value is known by the user node first.

mrshenli committed:
Update on "Make RRef::toHere() return a jit::Future"


This is step 1 to make PyRRef::toHere() non-blocking on caller.

Differential Revision: [D17936747](https://our.internmc.facebook.com/intern/diff/D17936747)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/rref.cpp:
```diff
@@ -131,32 +130,39 @@ const ForkId& UserRRef<T>::forkId() const {
+
+  fm->addCallback([future](const Message& message) {
+    RRefContext::handleException(message);
+    auto rfr = ScriptRRefFetchRet::fromMessage(message);
```
Actually, I am not sure if we would like to do that. `deserializeResponse` uses switch case on the message type to find the correct way to deserialize the response, but here we know for sure that the response should be  `ScriptRRefFetchRet` otherwise sth is wrong. Will it be reasonable if we directly call `ScriptRRefFetchRet::fromMessage` here to save the overhead of switch case?

xush6528 replied:
Makes sense. They are in pair.

Then, why do we need to add `SCRIPT_RREF_FETCH_RET` and `PYTHON_RREF_FETCH_RET` branches to `deserializeResponse `. Where is it used?

mrshenli replied:
Good catch!! They don't need to be in `deserializeResponse`. Let me remove them.

mrshenli committed:
Update on "Make RRef::toHere() return a jit::Future"


This is step 1 to make PyRRef::toHere() non-blocking on caller.

Differential Revision: [D17936747](https://our.internmc.facebook.com/intern/diff/D17936747)

[ghstack-poisoned]

xush6528 commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -71,8 +71,11 @@ std::unique_ptr<RpcCommandBase> deserializeResponse(const Message& response) {
+      return ScriptRRefFetchRet::fromMessage(response);
+    }
+    case MessageType::PYTHON_RREF_FETCH_RET: {
+      return PythonRRefFetchRet::fromMessage(response);
```
These 2 cases could be removed.

mrshenli committed:
Update on "Make RRef::toHere() return a jit::Future"


This is step 1 to make PyRRef::toHere() non-blocking on caller.

Differential Revision: [D17936747](https://our.internmc.facebook.com/intern/diff/D17936747)

[ghstack-poisoned]

pietern commented on torch/csrc/distributed/rpc/message.h:
```diff
@@ -24,23 +24,24 @@ enum MessageType {
-  CLEANUP_AUTOGRAD_CONTEXT_REQ = 18,
-  CLEANUP_AUTOGRAD_CONTEXT_RESP = 19,
+  CLEANUP_AUTOGRAD_CONTEXT_REQ = 19,
+  CLEANUP_AUTOGRAD_CONTEXT_RESP = 20,
```
If there is not a tight constraint on the maximum message ID (e.g. int8) then you could offset these categories by multiples of 100 or so to make adding new types a bit easier.

pietern commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -103,8 +103,7 @@ Message RequestCallbackImpl::processRpc(
           ctx.getOrCreateOwnerRRef<IValue>(srf.rrefId());
-      return std::move(RRefFetchRet(RRefFetchRet({rref->getValue()})))
-          .toMessage();
+      return std::move(ScriptRRefFetchRet({rref->getValue()})).toMessage();
```
Is std::move still needed on a temporary? IIRC a temporary qualifies as an rvalue.

edit: I checked and it should work without std::move (per godbolt): https://godbolt.org/z/-eeLnT

mrshenli replied:
Will address this on followup PRs. Thanks!

pietern commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -114,8 +113,7 @@ Message RequestCallbackImpl::processRpc(
           PythonRpcHandler::getInstance().serialize(rref->getValue());
-      return std::move(RRefFetchRet(RRefFetchRet(result.toIValues())))
-          .toMessage();
+      return std::move(PythonRRefFetchRet(result.toIValues())).toMessage();
```
Same as above.
